File: PixelSolution/PixelEngine/PVector3.cpp
#include "pch.h"
#include "PVector3.h"
#include <cmath>

PVector3 PVector3_Normalize(PVector3 vector3)
{
	const float lengthSquared = vector3.x * vector3.x
		+ vector3.y * vector3.y
		+ vector3.z * vector3.z;

	if (lengthSquared <= 1.0e-12f)
	{
		return PVector3(0.0f, 0.0f, 0.0f);
	}

	const float inverseLength = 1.0f / std::sqrt(lengthSquared);
	return PVector3(
		vector3.x * inverseLength,
		vector3.y * inverseLength,
		vector3.z * inverseLength);
}

float PVector3_Length(PVector3 vector3)
{
	return std::sqrt(vector3.x * vector3.x
		+ vector3.y * vector3.y
		+ vector3.z * vector3.z);
}

float PVector3_Distance(PVector3 from, PVector3 to)
{
	const float x = to.x - from.x;
	const float y = to.y - from.y;
	const float z = to.z - from.z;
	return std::sqrt(x * x + y * y + z * z);
}

float PVector3_Dot(PVector3 lhs, PVector3 rhs)
{
	return lhs.x * rhs.x + lhs.y * rhs.y + lhs.z * rhs.z;
}

PVector3 PVector3_Cross(PVector3 lhs, PVector3 rhs)
{
	return PVector3(
		lhs.y * rhs.z - lhs.z * rhs.y,
		lhs.z * rhs.x - lhs.x * rhs.z,
		lhs.x * rhs.y - lhs.y * rhs.x);
}

PVector3 PVector3_Direction(PVector3 from, PVector3 to)
{
	return PVector3_Normalize(PVector3(
		to.x - from.x,
		to.y - from.y,
		to.z - from.z));
}

PVector3 PVector3_Lerp(PVector3 from, PVector3 to, float amount)
{
	return PVector3(
		from.x + (to.x - from.x) * amount,
		from.y + (to.y - from.y) * amount,
		from.z + (to.z - from.z) * amount);
}

bool PVector3_ApproximatelyEquals(PVector3 lhs, PVector3 rhs, float epsilon)
{
	if (epsilon < 0.0f)
	{
		epsilon = -epsilon;
	}

	return std::fabs(lhs.x - rhs.x) <= epsilon
		&& std::fabs(lhs.y - rhs.y) <= epsilon
		&& std::fabs(lhs.z - rhs.z) <= epsilon;
}

PVector3 PVector3_Create(float x, float y, float z)
{
	return PVector3(x, y, z);
}

File: Survival/Engine/PVector3.lua
local ffi = require("ffi")
local dll = ffi.load("PixelEngine")

---@class PVector3
---@field x number
---@field y number
---@field z number
---@operator add(PVector3): PVector3
---@operator sub(PVector3): PVector3
---@operator mul(number|PVector3): PVector3
---@operator eq(PVector3): boolean
---@operator lt(PVector3): boolean
---@operator le(PVector3): boolean
---@field Normalize fun(self: PVector3): PVector3
---@field Length fun(self: PVector3): number
---@field Distance fun(self: PVector3, to: PVector3): number
---@field Dot fun(self: PVector3, rhs: PVector3): number
---@field Cross fun(self: PVector3, rhs: PVector3): PVector3
---@field Direction fun(self: PVector3, to: PVector3): PVector3
---@field Lerp fun(self: PVector3, to: PVector3, amount: number): PVector3
---@field ApproximatelyEquals fun(self: PVector3, rhs: PVector3, epsilon: number): boolean
ffi.cdef[[
	typedef struct {
		float x;
		float y;
		float z;
	} PVector3;

	PVector3 PVector3_Normalize(PVector3 normalizeVector3);
	float PVector3_Length(PVector3 vector3);
	float PVector3_Distance(PVector3 from,PVector3 to);
	float PVector3_Dot(PVector3 lhs,PVector3 rhs);
	PVector3 PVector3_Cross(PVector3 lhs,PVector3 rhs);
	PVector3 PVector3_Direction(PVector3 from,PVector3 to);
	PVector3 PVector3_Lerp(PVector3 from,PVector3 to,float amount);
	bool PVector3_ApproximatelyEquals(PVector3 lhs,PVector3 rhs,float epsilon);
	PVector3 PVector3_Create(float x,float y,float z);
]]

local PVector3_value_mt = {
	__index = {
		Normalize = function(normalizeVector3)
			return dll.PVector3_Normalize(normalizeVector3)
		end,
		Length = function(vector3)
			return dll.PVector3_Length(vector3)
		end,
		Distance = function(from,to)
			return dll.PVector3_Distance(from,to)
		end,
		Dot = function(lhs,rhs)
			return dll.PVector3_Dot(lhs,rhs)
		end,
		Cross = function(lhs,rhs)
			return dll.PVector3_Cross(lhs,rhs)
		end,
		Direction = function(from,to)
			return dll.PVector3_Direction(from,to)
		end,
		Lerp = function(from,to,amount)
			return dll.PVector3_Lerp(from,to,amount)
		end,
		ApproximatelyEquals = function(lhs,rhs,epsilon)
			return dll.PVector3_ApproximatelyEquals(lhs,rhs,epsilon)
		end,
	},
	__eq = function(a, b)
		return a.x == b.x and a.y == b.y and a.z == b.z
	end,
	__lt = function(a, b)
		return a.x < b.x and a.y < b.y and a.z < b.z
	end,
	__le = function(a, b)
		return a.x <= b.x and a.y <= b.y and a.z <= b.z
	end,
	__add = function(a, b)
		return dll.PVector3_Create(a.x + b.x, a.y + b.y, a.z + b.z)
	end,
	__sub = function(a, b)
		return dll.PVector3_Create(a.x - b.x, a.y - b.y, a.z - b.z)
	end,
	__mul = function(a, b)
		if type(a) == "number" then
			return dll.PVector3_Create(a * b.x, a * b.y, a * b.z)
		end
		if type(b) == "number" then
			return dll.PVector3_Create(a.x * b, a.y * b, a.z * b)
		end
		return dll.PVector3_Create(a.x * b.x, a.y * b.y, a.z * b.z)
	end
}

ffi.metatype("PVector3", PVector3_value_mt)

---@class PVector3
PVector3 = PVector3 or {}
setmetatable(PVector3, {
	__call = function(_,x,y,z)
		return dll.PVector3_Create(x,y,z)
	end
})

return PVector3